linc_convert/utils/io/zarr/drivers/tensorstore.py

```diff
@@ -0,0 +1,737 @@
+"""TensorStore driver for Zarr arrays and groups."""
+
+import json
+import math
+import os
+from numbers import Number
+from os import PathLike
+from typing import (
+    Any,
+    Iterator,
+    Literal,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    Unpack,
+    )
+from urllib.parse import urlparse
+
+import numcodecs
+import numcodecs.abc
+import numpy as np
+import tensorstore as ts
+import zarr
+from numpy.typing import ArrayLike, DTypeLike
+from upath import UPath
+
+from linc_convert.utils.io.zarr.abc import ZarrArray, ZarrArrayConfig, ZarrGroup
+from linc_convert.utils.zarr_config import ZarrConfig
+
+
+class ZarrTSArray(ZarrArray):
+    """Zarr array backed by TensorStore."""
+
+    def __init__(self, ts_array: ts.TensorStore) -> None:
+        """
+        Initialize the ZarrTSArray with a TensorStore array.
+
+        Parameters
+        ----------
+        ts_array : tensorstore.TensorStore
+            Underlying TensorStore array.
+        """
+        super().__init__(str(ts_array.kvstore.path))
+        self._ts = ts_array
+
+    @property
+    def ndim(self) -> int:
+        """Number of dimensions of the array."""
+        return self._ts.ndim
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        """Shape of the array."""
+        return self._ts.shape
+
+    @property
+    def dtype(self) -> np.dtype:
+        """Data type of the array."""
+        return self._ts.dtype.numpy_dtype
+
+    @property
+    def chunks(self) -> Tuple[int, ...]:
+        """Chunk shape for the array."""
+        return self._ts.chunk_layout.read_chunk.shape
+
+    @property
+    def shards(self) -> Optional[Tuple[int, ...]]:
+        """Shard shape, if supported; otherwise None."""
+        read_shape = self._ts.chunk_layout.read_chunk.shape
+        write_shape = self._ts.chunk_layout.write_chunk.shape
+        return None if read_shape == write_shape else write_shape
+
+    @property
+    def attrs(self) -> Mapping[str, Any]:
+        """Access metadata/attributes for this node."""
+        # TODO: TensorStore currently doesn’t expose arbitrary attrs, so return empty.
+        return {}
```

Comment:
Can we load them ourselves, using fsspec?
We know the path to the store and the zarr version, so can also guess the path to the attr file.